src/chainstate/stacks/miner.rs

```diff
@@ -1139,34 +1123,30 @@ impl StacksBlockBuilder {
         }
     }
 
-    /// Begin mining an epoch's transactions.
-    /// Returns an open ClarityTx for mining the block, as well as the ExecutionCost of any confirmed
-    ///  microblocks.
-    /// NOTE: even though we don't yet know the block hash, the Clarity VM ensures that a
-    /// transaction can't query information about the _current_ block (i.e. information that is not
-    /// yet known).
-    pub fn epoch_begin<'a>(
+    /// This function should be called before `epoch_begin`.
+    /// It loads the parent microblock stream, sets the parent microblock, and returns
+    /// data necessary for `epoch_begin`.
+    /// Returns chainstate transaction, clarity instance, burnchain header hash
+    /// of the burn tip, burn tip height + 1, the parent microblock stream,
+    /// the parent consensus hash, the parent header hash, and a bool
+    /// representing whether the network is mainnet or not.
+    pub fn pre_epoch_begin<'a>(
```

Comment:
I am pretty sure that this is necessary for the borrow-checker/ownership checks: something higher than the scope of `epoch_begin` needs to own the returned `ChainstateTx` and `ClarityInstance` (returned by `chainstate.chainstate_tx_begin()?;`), so this code cannot all live in `epoch_begin`, and that line cannot be called by itself, because a lot of the setup work in `pre_epoch_begin` needs reference the `StacksChainState`, which cannot be used while the `ChainstateTx` and `ClarityInstance` are alive (because their lifetimes are bound to a mutable reference to the `StacksChainState`.

Reply:
It seems like it should be possible to expand the definition of `epoch_begin` to allow all of this.

But, I also don't feel that strongly, so I am OK with it.

Reply:
> I am pretty sure that this is necessary for the borrow-checker/ownership checks:

Yeah I see that now.

@pavitthrap Can we call this something more informative, in order to communicate this?  Perhaps `epoch_setup()`?